### DISCUSSION TITLE
Private CircleCI orbs

### DISCUSSION THREAD

**Discussion by nm-morais** [renovatebot/renovate#15438] [Request Help] [answered]
Hey folks, we are trying to have renovate auto-update our circleCI orbs that are private.

I've taken a look at the [docs](https://docs.renovatebot.com/modules/manager/circleci/), but I can't find any relevant information regarding private orbs, does anyone know if this is possible? If so, could you point me to some docs for this?

I've tried setting this block on "hostRules", but It does not seem to work.

```
   {
      "abortOnError": true,
      "enabled": true,
      "matchHost": "circleci.com",
      "username": "***",
      "encrypted": {
        "token": "****"
      }
    }
```

Cheers

**Comment by rarkins** [comment]
Have you been able to test your credentials manually, e.g. using curl?

Normally you should have either a token or a username/password.

**Reply by nm-morais** [reply]
Thanks for the quick reply :D 

I have been able to test the credentials using curl, circleCI accepts 2 types of auth using its API token:

 - sending a header with "Circle-Token: $CIRCLECI_TOKEN" -> haven't tested this one because AFAIK renovate does not support custom authentication headers
 
 - sending the API token as the username for HTTP Basic Authentication -> This works with `curl https://circleci.com/api/v2/me -u "$CIRCLE_TOKEN"  -v`, and setting an empty password

These are the two only ways I've been able to authenticate using curl.

Now, to replicate this behavior using renovate, I've tried setting this configuration:

``` 
 {
      "abortOnError": true,
      "enabled": true,
      "matchHost": "circleci.com",
      "authType": "Basic",
      "encrypted": {
        "username": "xxxxx"
      },
      "password": ""
    }
```

But still getting `Failed to look up dependency` error on renovate dashboard.. 
What could be wrong here?

**Reply by rarkins** [reply]
Is it possible in some verbose mode to see how `curl` translates the `-u` and missing password into headers?

**Reply by nm-morais** [reply]
curl uses b64 to encode the password/username into a pair, split using ":".

So, if I curl some url using `curl <some-website> -u "abc"  -v` and provide an empty password, It produces the header: `Authorization: Basic YWJjOg==`

And if i run `echo "YWJjOg==" | base64 -d`, this returns -> `abc:`

**Reply by rarkins** [reply]
That *should* be what Renovate does too. We'll need to do some code inspection

**Reply by nm-morais** [reply]
Is there a way to increase the log verbosity on github-hosted renovate? Might help debug what is being sent to circle...

Please let me know if you find anything

**Reply by rarkins** [reply]
We're a bit careful about logging secrets and authentication to the potential for accidental leaks.

Can you create a public reproduction I can debug with?

- Use a made-up orb name
- Don't encrypt the username, just put in something dummy like `abc123`

BTW in your host rule, drop `authType=Basic` - it should be implied automatically when there's username/password. Keep it as simple as possible.

**Reply by nm-morais** [reply]
Thanks for the feedback :D 

I've created a repo [here](https://github.com/nm-morais/renovate-circleci-private-orb-demo) 

Thanks for the help

**Reply by rarkins** [reply]
I noticed that we do a GraphQL POST to a /graphql-unstable endpoint and trying to verify that it's the latest/correct one and that it supports private orbs. Have you tested curl to *that* endpoint or used an alternative for querying your orb? If you only queried for your user endpoint then it doesn't prove that the token you have necessarily works for the orb you need.

**Reply by nm-morais** [reply]
Hey rarkins! 

I have tested multiple queries to Circle's API, and I think the problem is that the authentication methods supported for REST are different form those supported from graphQL.

I have tried the following authentication headers for the GraphQL endpoint (querying the orbs):

- `Authorization: <API-token>` -> this works as expected
- `Circle-Token: <API-token>` -> doesn't work
- `Authorization: Bearer  <API-token>` -> doesn't work
- `Authorization: Basic <base64 representation of '<API-token>:'` -> doesn't work

It's also worth mentioning that I managed to extract the latest version information for a private orb using the first authentication method, so the key I've been testing with has all the necessary permissions for querying orb versions.

Given this, do you know if it is possible to reproduce the behavior of the first authentication method using Renovate? 

Thanks for all the help so far 🙂

**Reply by rarkins** [reply]
Can you try a host rule like this?

```
   {
      "matchHost": "circleci.com",
      "authType": "Token-Only",
      "encrypted": {
        "token": "****"
      }
    }
```

Encrypt the API token and put it into that encrypted `token` field. This config hopefully strips away the Bearer/Basic prefix in the `authorization` header.

**Reply by nm-morais** [reply]
That works! Thank you so much

**Reply by rarkins** [reply]
Great! Thank you for your help resolving it. Would you like to submit a PR to the lib/modules/manager/circleci/readme.md file describing it? It will get included in our docs

**Comment by HonkingGoose** [comment, marked as answer]
I opened a PR to put the information in our docs:

- #15939
